src/pkg/utils/oras.go

```diff
@@ -90,37 +91,84 @@ func (o *OrasRemote) withAuthClient(ref registry.Reference) (*auth.Client, error
 }
 
 // Transport is an http.RoundTripper that keeps track of the in-flight
-// request and add hooks to report HTTP tracing events.
+// request and add hooks to report upload progress.
 type Transport struct {
-	http.RoundTripper
-	orasRemote *OrasRemote
+	Base       http.RoundTripper
+	OrasRemote *OrasRemote
+	Policy     retry.Policy
 }
 
 // NewTransport returns a custom transport that tracks an http.RoundTripper and an OrasRemote reference.
 func NewTransport(base http.RoundTripper, o *OrasRemote) *Transport {
-	return &Transport{base, o}
+	return &Transport{
+		Base:       base,
+		OrasRemote: o,
+	}
 }
 
-type readCloser struct {
-	io.Reader
-	io.Closer
+// RoundTrip is mirrored from retry, but instead of calling retry's private t.roundTrip(), this uses
+// our own which has interactions w/ message.ProgressBar
+func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx := req.Context()
+	policy := retry.DefaultPolicy
+	attempt := 0
+	for {
+		resp, respErr := t.roundTrip(req)
+		duration, err := policy.Retry(attempt, resp, respErr)
+		if err != nil {
+			if respErr == nil {
+				resp.Body.Close()
+			}
+			return nil, err
+		}
+		if duration < 0 {
+			return resp, respErr
+		}
+
+		// rewind the body if possible
+		if req.Body != nil {
+			if req.GetBody == nil {
+				// body can't be rewound, so we can't retry
+				return resp, respErr
+			}
+			body, err := req.GetBody()
+			if err != nil {
+				// failed to rewind the body, so we can't retry
+				return resp, respErr
+			}
+			req.Body = body
+		}
+
+		// close the response body if needed
+		if respErr == nil {
+			resp.Body.Close()
+		}
+
+		timer := time.NewTimer(duration)
+		select {
+		case <-ctx.Done():
+			timer.Stop()
+			return nil, ctx.Err()
+		case <-timer.C:
+		}
+		attempt++
+	}
```

Comment:
Can this be moved to a better file (since it isn't oras specific and could be generally useful?

Also can you link to where this was from